Support "diagonal" primitives with no/slow JVP batch rule

jpbrodrick89 opened the pull request:
BREAKING CHANGE: Diagonals are no longer "extracted" from operators but rely on the promise of the `diagonal` tag being accurate and fulfilled. If this promise is broken solutions may differ from user expectations.

### Preface

At present, both `JacobianLinearOperator` and `FunctionLinearOperator` require full materialisation even if provided with a `diagonal` tag. This seems self-evidently expensive (in practice it certainly can be but more often is not, see below) and requires the underlying function (which could potentially be a custom primitive) to have a batching rule. As it is currently the case that tags are considered to be a "promise" and are unchecked with no guarantee of behaviour, there are some shortcuts we can take.

### Changes made

The proposal here is to use the observation that the diagonal of a matrix can be obtained by pre/post-multiplying it by a unit vector and thereby re-write the single-dispatch `diagonal` method for `JacobianLinearOperator` and `FunctionLinearOperator` so that the `as_matrix()` method is not required. For `JacobianLinearOperator` either `jax.jvp` or `jax.vjp` will be called depending on the `jac` keyword (forward-mode should always be more efficient but meeting the user's expectation will avoid issues if forward-mode is not supported such as when using a `custom_vjp` is used). For `FunctionLinearOperator`, we can just use `self.mv`.

However, if the matrix is not actually diagonal this identity will not hold and results may be unexpected due to contributions from off-diagonals.

I considered using `operator.transpose().mv` instead of writing out `vjp` but if the matrix is tagged as `symmetric` then this would end up calling `jacrev` instead of `vjp`.

### Why is this helpful?

When using `lineax` directly one can of course just define a `DiagonalOperator` instead of a more general `JacobianLinearOperator`, but this is not always possible. For example, when using `optimistix`, the operator is instantiated within the optimisation routine and the only way to inform the optimiser about the underlying structure of the matrix is through `tags`. Therefore, if the function being optimised is a primitive (e.g. an FFI) with a JVP rule that does not support batching a user is stuck. If a slow batching rule, such as `vmap_method="sequential"`, is used the current approach is also painfully slow for large matrix sizes.

### Performance impact

I had initially hoped this to have a minor positive impact on performance across the board, but as ever I have massively underestimated the power of XLA. In practice, whether this PR seems to improve performance (e.g. for a `linear_solve` or an `optimistix.root_find`) of a _pure jax function_ appears to fluctuate with array size. By playing around with different `XLA_FLAGS` and other environment variables, my best guess is that this is mostly due to threading; a `vmap` applied to a `jnp.eye` is threaded much more aggressively meaning that the apparent time complexity appears to be of lower order than a more direct approach. However, when I tried to eliminate threading this PR still seems to have an 8–10% negative impact on performance for array sizes > 100 on an `optimistix.root_find`.

<details>
<summary>
Pure `jax` comparison: using `jvp` when attempting to enfore single-threadedness is about 14µs faster.
</summary>

```python
@jax.jit
def from_eye(x):
    eye = jnp.eye(len(x), dtype=x.dtype)
    jac = jax.vmap(lambda t: jnp.cos(x)*t, out_axes=-1)(eye)
    return jnp.diag(jac)

@jax.jit
def direct(x):
    return jnp.cos(x)
```
It seems self-evident that the second function should be more efficient, however with the new `thunk` runtime on my Mac `from_eye` runs faster than `direct` (referred to as `wrapped` in the diagram below, you can ignore `unwrapped` and `vmap` as similar performance) for array sizes > ~1.5E4:
![image](https://github.com/user-attachments/assets/1ebf68bb-b08c-453c-89dd-dcf589309ff1)

Disabling the `thunk` runtime (with `XLA_FLAGS=--xla_cpu_use_thunk_runtime=false` which is reported to run faster in some circumstances) decreases the gap between the two by slightly slowing down the `eye` implementation and accelerating the `direct` approach:
![image](https://github.com/user-attachments/assets/2e4ec841-3b01-48e8-bbd7-984bd82824a4)

Going further and following all suggestions in [github.com/jax-ml/jax/discussions/22739](https://github.com/jax-ml/jax/discussions/22739) to limit to one thread/core and we can see the `direct` approach is now consistently about 14µs faster:
![image](https://github.com/user-attachments/assets/114dda54-cced-4a45-992d-a46361c29634)
</details>

<details>
<summary>
`linear_solve` significantly faster (often >2x) for array sizes <2E4 using thunk runtime, but runs about 6–10% slower for large array sizes when disabling and attempting to enforce a single thread
</summary>

<details>
<summary>
Code tested
</summary>

```python
solver = lineax.Diagonal(well_posed=True)

def double(x, args):
    return (2.0 * jnp.ones_like(x)) * x

@jax.jit
def jac_op_to_sol(rhs):
    op = lineax.JacobianLinearOperator(double, rhs, tags=frozenset({lineax.diagonal_tag}))
    return lineax.linear_solve(op, rhs, solver, throw=False)

@jax.jit
def func_op_to_sol(rhs):
    op = lineax.FunctionLinearOperator(lambda x: double(x, None), jax.eval_shape(lambda: rhs), tags=frozenset({lineax.diagonal_tag}))
    return lineax.linear_solve(op, rhs, solver, throw=False)
```
</details>

Using standard thunk runtime and `EQX_ON_ERROR=nan` we see significant speedup for array sizes < 1E4
![image](https://github.com/user-attachments/assets/67f7f695-7c33-4edb-a117-91b4ae13b69f)

Enforcing single-thread the performance between the old and the new approaches is very similar but tracks at about 6–10% slower for larger array sizes.
![image](https://github.com/user-attachments/assets/44fa8676-3994-4c70-a323-57c69661f8ef)

(Note that `DiagonalOperator` is actually slower somehow.)

</details>

<details>
<summary>
Similar behaviour is observed with `optimistix.root_find` (but with more modest gains, and some hits for larger array sizes)
</summary>

I compared performance for a multi-root find of the `sin` function (with `EQX_ON_ERROR=nan`):
```python
x = jnp.repeat(jnp.linspace(-10., 10., num=20), n // 20)
newton = optx.Newton(rtol=1e-8, atol=1e-8, linear_solver=lx.Diagonal())
optx.root_find(lambda x, args: jnp.sin(x), newton, x, max_steps=8, tags=frozenset({lx.diagonal_tag}), throw=False).value
```

Default settings (`jax` 0.6.1, `main` vs this branch of `lineax) with and without standard `thunk` runtime:
![image](https://github.com/user-attachments/assets/0dc344f4-588e-485a-b4a0-20949617bbc4)

In both runtimes this PR improves/maintains performance by up to a factor of 2 for arrays of size up to 1E4 at which point it becomes slightly slow than the current version (by ~8%).

However, limiting to one thread as best as I can most of the noise is eliminated and the two have very similar performance time (the change tracking about 6% slower) except for an array size of 20 where the proposed change is faster:
![image](https://github.com/user-attachments/assets/cf66df23-36e1-4655-aa37-212a5ef49908)

</details>

<details>
<summary>
Much more substantial performance improvement (8x or higher) is observed for primitives that only support `sequential` batching rules
</summary>

This is a very contrived example, but based on very real use cases we have over at [tesseract-core](https://github.com/pasteurlabs/tesseract-core) and [tesseract-jax](https://github.com/pasteurlabs/tesseract-jax). I have defined a new primitive version of `sin` with a `jvp` rule that batches sequentially and is therefore slow and doesn't benefit from compilation/threading in the same way:

<details>
<summary>
Code for primitives
</summary>

```python
import numpy as onp

import jax
import jax.numpy as jnp
from jax.interpreters import ad, batching, mlir
from jax.core import ShapedArray
from jax.extend import core

from jax._src.lib.mlir.dialects import hlo
from jax._src.ffi import ffi_batching_rule

from functools import partial

jax.config.update("jax_enable_x64", True)

sin_p = core.Primitive("mysin")
cos_mult_p = core.Primitive("cos_mult")

def sin_prim(x):
    return sin_p.bind(x)

def cos_mult_prim(*args):
    p = args[0]
    return cos_mult_p.bind(*args,
                           vmap_method="sequential",
                           result_avals=(ShapedArray(p.shape, p.dtype), ))

def cos_mult_impl(*args, vmap_method="sequential", result_avals=None):
    del vmap_method, result_avals
    return onp.cos(args[0]) * args[1]

def cos_mult_abstract_eval(*args, vmap_method="sequential", result_avals=None):
    del vmap_method, result_avals
    return ShapedArray(args[0].shape, args[0].dtype)

def sin_jvp(p, t):
    return sin_prim(p[0]), cos_mult_prim(p[0], t[0])

def sin_lowering(ctx, x):
    return [ hlo.SineOp(x).result ]

def cos_mult_lowering(ctx, p, t, *, vmap_method="sequential", result_avals=None):
    del vmap_method, result_avals
    return [ hlo.MulOp(hlo.CosineOp(p), t).result ]

def cos_mult_batch(args, axes, vmap_method="sequential", result_avals=None):
    as_tup, out_dims =  ffi_batching_rule(cos_mult_p, args, axes, 
                            vmap_method="sequential",
                            result_avals=(ShapedArray(args[0].shape, args[0].dtype),))
    return jnp.stack(list(as_tup)), out_dims[0]

sin_p.def_impl(onp.sin)
cos_mult_p.def_impl(onp.cos)
sin_p.def_abstract_eval(lambda x: ShapedArray(x.shape, x.dtype))
cos_mult_p.def_abstract_eval(cos_mult_abstract_eval)
mlir.register_lowering(sin_p, sin_lowering)
mlir.register_lowering(cos_mult_p, cos_mult_lowering)
ad.primitive_jvps[sin_p] = sin_jvp
# If the `cos_mult_p` batching rule isn't present this won't run on the current version
# but will with this PR
batching.primitive_batchers[cos_mult_p] = cos_mult_batch
```
</details>

I then ran the same tests as before but with `sin_p` instead of `jnp.sin` and we can see the time complexity of the current version is almost quadratic for array sizes greater than 100 (as one would naively expect for a dense jacobian) meaning that speedups range from a factor of 2 (array size of 20) to a factor of  8 (array size of 5000) and higher:
![image](https://github.com/user-attachments/assets/0e778ea2-8a19-4c33-b1e8-9ea2cd6912ba)

</details>
 
~~Running `benchmarks/solver_speed.py` shows a negligible improvement in the single `Diagonal` solve but a 50% faster batch solve, this could of course be down to noise as the solve is only timed once.~~ (This uses `lx.Diagonal` so not relevant and probably just a fluke.)

### Testing done
- [x] CI passes after modifying `test_diagonal` such that operators are actually initialised with diagonal matrices
- [x] Can find root (using both `Newton` and `Bisection`) of scalar function with no batching rule and take gradients through the root solve (not possible previously) this tests both `JacobianLinearOperator` and `FunctionLinearOperator` in action
- [x] Can obtain `diagonal` from `JacobianLinearOperator` with `jac="bwd"` 

Happy to perform any further requested testing you see fit/necessary. I appreciate I haven't managed to test reverse-mode especially extensively.

### Next steps

In a future PR, I would like to do something similar for other structures (e.g. tridiagonal) this should address the large O(n) discrepancy observed in #149 (but not the O(0) discrepancy). I believe this will be a much more consistent and meaningful gain than observed here. This PR here should likely be a lot easier to grok and reason about the concept and discuss framework/design choices (although maybe not the performance impact :) ) before building out further.

jpbrodrick89 committed:
Merge pull request #1 from patrick-kidger/main

merge main into fork

jpbrodrick89 committed:
Streamline diagonalisation by avoiding computation/instantiation of entire Jacobian matrix

jpbrodrick89 committed:
changed to using mv instead of jvp

jpbrodrick89 committed:
initialise test_diagonal  operators with diagonal matrix

jpbrodrick89 committed:
Merge branch 'patrick-kidger:main' into main

jpbrodrick89 committed:
remove redundant original diagonal overload for  FunctionLinearOperator

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
unravel instead of map

patrick-kidger commented on lineax/_operator.py:
```diff
@@ -1363,10 +1363,40 @@ def diagonal(operator: AbstractLinearOperator) -> Shaped[Array, " size"]:
 
 @diagonal.register(MatrixLinearOperator)
 @diagonal.register(PyTreeLinearOperator)
```
Side note, I think we could make the `PyTreeLinearOperator` case a little more efficient by calling `jnp.diag` on each 'diagonal leaf', and concatenate those together. (You don't have to do that here, just spotting it.)

patrick-kidger commented on lineax/_operator.py:
```diff
@@ -1363,10 +1363,40 @@ def diagonal(operator: AbstractLinearOperator) -> Shaped[Array, " size"]:
 @diagonal.register(JacobianLinearOperator)
+def _(operator):
+    with jax.ensure_compile_time_eval():
+        flat, unravel = strip_weak_dtype(
+            eqx.filter_eval_shape(jfu.ravel_pytree, operator.in_structure())
+        )
+        basis = jnp.ones(flat.size, dtype=flat.dtype)
```
Could this just be `basis = jtu.tree_map(jnp.ones_like, operator.in_structure())`?

Either way, why is the compile-time-eval needed here?

jpbrodrick89 replied:
I used your suggestion in my original commit but this seems to run faster based on my testing, I think both compile time evaluation (which was inspired by sparsejac implementation) and a single ones call led to performance improvements but I can try and dig in/disambiguate a bit if interest.

patrick-kidger replied:
Hmm, curious. I could see the compile-time-eval tricking XLA into constant-propagating where it might otherwise choose not to. And perhaps having a single constant rather than n constants also leads to some kind of performance improvement somehow.

Happy to leave this as-is if this is apparently fastest.

patrick-kidger commented on lineax/_operator.py:
```diff
@@ -1363,10 +1363,40 @@ def diagonal(operator: AbstractLinearOperator) -> Shaped[Array, " size"]:
+    elif operator.jac == "bwd":
+        fn = _NoAuxOut(_NoAuxIn(operator.fn, operator.args))
+        _, vjp_fun = jax.vjp(fn, operator.x)
+        diag_as_pytree = vjp_fun(unravel(basis))
```
I think this will fail for operators with different input and output structures. (They might still be mathematically square, just split up into a pytree in different ways / with different dtypes.) This needs to be a basis formed from `operator.out_structure()`.

jpbrodrick89 replied:
I just added a test for mismatched input and output structure and looks like these are not allowed to be diagonal in the current version of lineax. Shall we leave relaxing that assumption and fixing this for future work? This implementation should work fine as long as the assumption holds.

```
FAILED tests/test_operator.py::test_is_symmetric[float64] - ValueError: Symmetric matrices must have matching input and output structures. Got input structure (ShapeDtypeStruct(shape=(2,), dtype=float64), ShapeDtypeStruct(shape=(), dtype=float64)) and output structure (ShapeDtypeStruct(shape=(), dty...
FAILED tests/test_operator.py::test_is_symmetric[complex128] - ValueError: Symmetric matrices must have matching input and output structures. Got input structure (ShapeDtypeStruct(shape=(2,), dtype=complex128), ShapeDtypeStruct(shape=(), dtype=complex128)) and output structure (ShapeDtypeStruct(shape=(...
FAILED tests/test_operator.py::test_is_diagonal[float64] - ValueError: Symmetric matrices must have matching input and output structures. Got input structure (ShapeDtypeStruct(shape=(2,), dtype=float64), ShapeDtypeStruct(shape=(), dtype=float64)) and output structure (ShapeDtypeStruct(shape=(), dty...
FAILED tests/test_operator.py::test_is_diagonal[complex128] - ValueError: Symmetric matrices must have matching input and output structures. Got input structure (ShapeDtypeStruct(shape=(2,), dtype=complex128), ShapeDtypeStruct(shape=(), dtype=complex128)) and output structure (ShapeDtypeStruct(shape=(...
```

jpbrodrick89 replied:
Never mind, these are `PyTreeLinearOperators` but I need to address for `JacobianLinearOperators`

patrick-kidger reviewed (commented):
Nice, I really like seeing a colouring approach like this!

I will note however that `lx.diagonal` is documented as `Extracts the diagonal from a linear operator, and returns a vector`, which is meant to include extracting the diagonal from nondiagonal operators. I think the implementations you have here should check with `is_diagonal` to determine whether to dispatch to the new or the old implementation.

patrick-kidger commented:
> Happy to address these during the week, just wanted to check: I did a quick trawl through the code and I don't think diagonal or tridagonal is ever used by operators that don't have the corresponding tag, so we have an option to just change the doc string instead of you like?

However they're actually a standalone public API themselves, which could be used independent of the solvers. :)

jpbrodrick89 committed:
continue to extract diagonal if no diagonal tag

jpbrodrick89 committed:
rmeove print statement in test_operator

jpbrodrick89 commented:
Sorry for abandoning this for so long as the day job took over. I returned as it was found to provide orders of magnitude impact on a problem I was working (root finding over multiple interpolations) even for small array sizes (200–2000). I have addressed the main point of retaining extraction of the diagonal when the diagonal tag is missing. However, your other two comments seem at odds with each other: either we ensure input/output structures match (which is actually the case, see above) enabling us to extract diagonal leafs for `PyTreeLinearOperator` and leaving the `JacobianLinearOperator` vjp implementation as it is, or we relax this assumption and are no longer able to extract diagonal leafs as there is not guaranteed to be such an object.

Commit:
Decrease default value to prevent overflow in 32-bit.

Commit:
use maximum value of runtime dtype

Commit:
limit condition numbers for sensitive, iterative solvers

Commit:
cond_cutoff -> 900 for GMRES, LSMR

Commit:
decrease condition number cutoff for LSMR

Commit:
implement safeguard for int overflow in 32 bit

Commit:
add extra safeguard for complex dtypes

patrick-kidger commented:
Sorry for the long delay getting back to you, some personal life things took over for a while.

So, now to actually answer your question: good point.

I imagine we could probably do the 'diagonal leaf' approach when the structures match, and go for the more expensive approach when they don't?

adconner committed:
Replace allow_dependent_rows and allow_dependent_columns with assume_full_rank (#158)

The two functions allow_dependent_{rows,columns} together did the job of
answering if the solver accepts full rank matrices for the purposes of
the jvp. Allowing them to be implemented separately created some issues:
1) Invalid states were representable. Eg. What does it mean that
   dependent columns are allowed for square matrices if dependent rows
   are not? What does it mean that dependent rows are not allowed for
   matrices with more rows than columns?
2) As the functions accept operator as input, a custom solver could in
   principle decide its answer based on operator's dynamic value rather
   than only jax compilation static information regarding it, as in all
   the lineax defined solvers. This would prevent jax compilation and
   jit.

Both issues are addressed by asking the solver to report only if it
assumes the input is numerically full rank. If this assumption is
exactly violated, its behavior is allowed to be undefined, and is
allowed to error, produce NaN values, and produce invalid values.

patrick-kidger committed:
pyright fixes

adconner committed:
Implement Normal, a solver applying an inner solver to the normal equations (#159)

patrick-kidger committed:
Tidy up results. (#170)

patrick-kidger committed:
Update infra (#182)

Commit:
fix typo

Commit:
reset condition number check

Commit:
simplify dtype conversion: reuse function from our misc module

johannahaffner committed:
Merge pull request #176 from johannahaffner/lsmr-init-overflow

Decrease default value to prevent overflow in 32-bit.

patrick-kidger committed:
Moved bool states to Static.

There seem to be some spurious downstream failures in Diffrax with JAX 0.8.2 otherwise. Probably JAX has started promoting these to tracers on some unusual codepath.

jpbrodrick89 committed:
simplify jac=bwd behaviour

jpbrodrick89 committed:
Merge branch 'dev' into jpb/diagfromjac

jpbrodrick89 committed:
add PyTreeLinearOperator optimisations for mv and diagonal

jpbrodrick89 committed:
add type annotation to _leaf_from_keypath

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
add efficient Matrix mv path, and Jacobian/Function materialise path

jpbrodrick89 committed:
add try sparse materialise helper (#3)

* Add sparse materialisation helper and efficient diagonal paths

This PR introduces _try_sparse_materialise helper and optimizes diagonal
operator handling throughout lineax.

Key changes:
- Add _try_sparse_materialise() that converts diagonal-tagged operators to
  DiagonalLinearOperator, preserving pytree structure via unravel
- Add efficient diagonal() for JLO/FLO using single JVP/VJP with ones basis
- Add efficient diagonal() for Composed: diag(A @ B) = diag(A) * diag(B)
- Simplify mv() for MLO, PTLO, Add, Composed to use _try_sparse_materialise
- Apply early sparse materialisation in materialise() registrations

Aux handling:
- Fix bug: linearise/materialise now preserve aux on AuxLinearOperator
- Preserve aux from first operator in Composed (output comes from op1)
- Inner aux in Add children silently stripped (unclear semantics - may
  warrant guards in future)

---------

Co-authored-by: jpbrodrick89 <jpbrodrick89@users.noreply.github.com>

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
unify JLO and FLO diagonal registrations

jpbrodrick89 committed:
rename sparse variable -> maybe_sparse_op

jpbrodrick89 committed:
move and shorten construct diagonal basis

jpbrodrick89 committed:
revert unecessary **2.0 -> more efficient **2

jpbrodrick89 committed:
comments on why we can't use try sparse materialise with JLO

jpbrodrick89 commented:
Note this is largely ready to go and includes #195 (but not #196), happy to get those in first and merge this in after for a cleaner diff.

Main change I've made is the `try_sparse_materialise` helper which converts an operator to `DiagonalLinearOperator` if it's diagonal, which can be used in `mv` or `materialise` and should make the code more maintainable and extensible (I shouldn't need to touch materialise in the tridiagonal PR except for JacobianLinearOperator. This is a bit of a breaking change though as materialise does not just return self for as many operators but instead materialises as the "most efficient" representation of the operator. Happy to hear any thoughts. :-)

Also happy to be targeting a dev branch if you prefer as these are not just "fixes", one just doesn't exist right now.

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
out -> maybe_sparse_op

jpbrodrick89 committed:
merge main without pre-materialising IdentityLO

jpbrodrick89 committed:
remove duplicated AddLinearOperator registration

patrick-kidger commented on lineax/_operator.py:
```diff
@@ -329,6 +332,14 @@ def __init__(self, value):
         self.value = value
 
 
+def _leaf_from_keypath(pytree: PyTree, keypath: jtu.KeyPath) -> Array:
```
Nit, can we move this function closer to its callsite (much further down in the file)?

patrick-kidger reviewed (commented):
Okay indeed final thoughts – I think this all looks pretty good to me! Minor comments below only.

Whilst I think this is fine and I am happy to merge this as-is, I am wondering if the many special-cases are indicative of the need for some other structure here. (E.g. maybe we should arrange for `materialise` to always start with `_try_sparse_materialise`, and then fall back to per-operator overloads?) In general I'm not sure what that is though.

I don't think I'd like to see too much more complexity here beyond what we have now, though, e.g. I'm a bit antsy about what the analogous tridigonal implementations will look like layered on top of this.

jpbrodrick89 commented on lineax/_operator.py:
```diff
@@ -1917,17 +2008,20 @@ def _(operator):
 @diagonal.register(ComposedLinearOperator)
 def _(operator):
+    if is_diagonal(operator):
+        return diagonal(operator.operator1) * diagonal(operator.operator2)
```
Oh wait you mean the converse might not be true, yes good catch theoretically you're right. But right now in lineax that's not true as we don't look for tags on the composed linear operator. So if it's not made of two diagonal operator it won't REGISTER as being diagonal so it's fine for now unless you want to change that behaviour.

jpbrodrick89 committed:
deduplicate AddLinearOperator check registrations

jpbrodrick89 committed:
move leaf_from_keypath

jpbrodrick89 committed:
robust forward compatible diagonal extraction from ComposedLinearOperator

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 commented:
Tests still pass locally I'm wondering if its due to RNG (i.e. random vec is sometimes in range of singular matrix and sometimes not)?

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac
